src/Workspaces/Core/Portable/Workspace/ProjectSystem/FileWatchedReferenceFactory.cs

```diff
@@ -4,127 +4,98 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.IO;
-using System.Linq;
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.ProjectSystem;
 
-internal sealed class FileWatchedPortableExecutableReferenceFactory
+internal sealed class FileWatchedReferenceFactory<TReference>
+    where TReference : class
```

Comment:
made generic, so we can share the majority of logic across PortableExecutableReference and AnalyzerReference